include/boost/fusion/container/vector/detail/as_vector.hpp

```diff
@@ -8,12 +8,64 @@
 #define FUSION_AS_VECTOR_11052014_1801
 
 #include <boost/fusion/support/config.hpp>
+#include <boost/fusion/container/vector/detail/config.hpp>
 
 ///////////////////////////////////////////////////////////////////////////////
 // Without variadics, we will use the PP version
 ///////////////////////////////////////////////////////////////////////////////
+#if !defined(BOOST_FUSION_HAS_VARIADIC_VECTOR)
 # include <boost/fusion/container/vector/detail/cpp03/as_vector.hpp>
+#else
 
+///////////////////////////////////////////////////////////////////////////////
+// C++11 interface
+///////////////////////////////////////////////////////////////////////////////
+#include <boost/fusion/support/detail/index_sequence.hpp>
+#include <boost/fusion/container/vector/vector.hpp>
+#include <boost/fusion/iterator/value_of.hpp>
+#include <boost/fusion/iterator/deref.hpp>
+#include <boost/fusion/iterator/advance.hpp>
+#include <cstddef>
+
+namespace boost { namespace fusion { namespace detail
+{
+BOOST_FUSION_BARRIER_BEGIN
+
+    template <typename Indices>
+    struct as_vector_impl;
+
+    template <std::size_t ...Indices>
+    struct as_vector_impl<index_sequence<Indices...> >
+    {
+        template <typename Iterator>
+        struct apply
+        {
+            typedef vector<
+                numbered_vector_tag<sizeof...(Indices)>
+              , typename result_of::value_of<
+                    typename result_of::advance_c<Iterator, Indices>::type
+                >::type...
+            > type;
+        };
+
+        template <typename Iterator>
+        BOOST_CONSTEXPR BOOST_FUSION_GPU_ENABLED
+        static typename apply<Iterator>::type
+        call(Iterator i)
```

Comment:
An implicitly deduced return type function has a placeholder as return type, forming something of an "incomplete" function declaration. The function declaration is only completed within the definition, once `return` is seen. Any use of such functions, even those in which a declaration would suffice, need to cause an instantiation in order to have a definition for it to "complete" the declaration.

**7.1.6.4 [dcl.spec.auto]/12**

Return type deduction for a function template with a placeholder in its declared type occurs when the definition is instantiated even if the function body contains a **return** statement with a non-type-dependent operand. [_Note:_ Therefore, any use of a specialization of the function template will cause an implicit instantiation. Any errors that arise from this instantiation are not in the immediate context of the function type and can result in the program being ill-formed.  —_end note_] [_Example_:

    template <class T> auto f(T t) { return t; }  // return type deduced at instantiation time
    typedef decltype(f(1)) fint_t;                // instantiates f<int> to deduce return type
    template<class T> auto f(T* t) { return *t; }
    void g() { int (*p)(int*) = &f; }             // instantiates both fs to determine return types,
                                                  // chooses second

 —_end example_]